File: e3_python_app/tests/test_seed_catalogue.py
"""Tests for the exact and annotated E3 seed catalogue."""

from __future__ import annotations

import duckdb
import pandas as pd
import pytest

from e3app.errors import AppError
from e3app.seed_catalogue import (
    build_seed_catalogue_query,
    collect_seed_catalogue,
    filter_seed_catalogue,
    seed_catalogue_capability,
)


@pytest.fixture
def seed_connection() -> duckdb.DuckDBPyConnection:
    """Create seed evidence with one exact sequence match."""
    connection = duckdb.connect(":memory:")
    connection.execute(
        "CREATE TABLE candidate_evidence("
        "cluster_id VARCHAR, matched_seed_ids_calculated VARCHAR, "
        "seed_protein_names VARCHAR, seed_categories VARCHAR, "
        "seed_review_statuses VARCHAR, seed_ubiquitin_go_statuses VARCHAR, "
        "seed_organisms VARCHAR)"
    )
    connection.execute(
        "INSERT INTO candidate_evidence VALUES "
        "('cluster_1', 'S1;S2', 'Seed one;Seed two', 'U-box', 'reviewed', "
        "'positive', 'Arabidopsis thaliana'), "
        "('cluster_2', 'S1', 'Seed one', 'U-box', 'reviewed', 'positive', "
        "'Arabidopsis thaliana')"
    )
    connection.execute(
        "CREATE TABLE known_e3_seeds("
        "seed_id VARCHAR, source_value VARCHAR, source_column VARCHAR, "
        "source_row INTEGER, source_path VARCHAR, seed_metadata_json VARCHAR)"
    )
    connection.execute(
        "INSERT INTO known_e3_seeds VALUES "
        "('S1', 'S1', 'accession', 2, 'seeds.tsv', "
        "'{\"protein_names\":\"Seed one\",\"e3_category\":\"U-box\","
        "\"organism\":\"Arabidopsis thaliana\",\"taxon_id\":\"3702\"}'), "
        "('S2', 'S2', 'accession', 3, 'seeds.tsv', "
        "'{\"protein_names\":\"Seed two\",\"category\":\"BTB\"}')"
    )
    connection.execute(
        "CREATE TABLE candidate_group_member_sequences("
        "parsed_accession VARCHAR, raw_identifier VARCHAR, species VARCHAR, "
        "protein_sequence VARCHAR, is_input_candidate BOOLEAN)"
    )
    connection.execute(
        "INSERT INTO candidate_group_member_sequences VALUES "
        "('S1', 'sp|S1|SEED_ONE', 'Arabidopsis_thaliana', 'MAAA', false), "
        "('OTHER', 'OTHER', 'Homo_sapiens', 'MBBB', false)"
    )
    try:
        yield connection
    finally:
        connection.close()


def test_seed_catalogue_collects_annotations_and_available_sequence(
    seed_connection: duckdb.DuckDBPyConnection,
) -> None:
    """Seed identifiers remain distinct and exact mapped sequence is retained."""
    capability = seed_catalogue_capability(connection=seed_connection)
    assert capability["available"]
    assert capability["mode"] == "authority"
    assert capability["relation"] == "known_e3_seeds"
    assert capability["sequence_available"]
    result = collect_seed_catalogue(
        connection=seed_connection,
        maximum_rows=10,
    )
    assert result["seed_id"].tolist() == ["S1", "S2"]
    first = result.loc[result["seed_id"] == "S1"].iloc[0]
    assert first["source_cluster_count"] == 2
    assert first["source_cluster_ids"] == "cluster_1;cluster_2"
    assert first["seed_protein_names"] == "Seed one"
    assert first["seed_category"] == "U-box"
    assert first["seed_taxon_id"] == "3702"
    assert first["associated_seed_protein_names"] == ""
    assert first["annotation_scope"] == "exact seed authority row"
    assert first["protein_sequence"] == "MAAA"
    assert first["protein_sequence_length"] == 4
    assert first["sequence_available"]
    second = result.loc[result["seed_id"] == "S2"].iloc[0]
    assert second["protein_sequence"] == ""
    assert not second["sequence_available"]


def test_seed_catalogue_falls_back_to_cluster_associated_annotations() -> None:
    """Older resources remain useful without overstating annotation scope."""
    connection = duckdb.connect(":memory:")
    try:
        connection.execute(
            "CREATE TABLE candidate_evidence("
            "cluster_id VARCHAR, matched_seed_ids_calculated VARCHAR, "
            "seed_protein_names VARCHAR)"
        )
        connection.execute(
            "INSERT INTO candidate_evidence VALUES "
            "('cluster_1', 'S1;S2', 'Seed one;Seed two'), "
            "('cluster_2', 'S1', 'Seed one')"
        )
        capability = seed_catalogue_capability(connection=connection)
        assert capability["mode"] == "cluster_summary"
        result = collect_seed_catalogue(
            connection=connection,
            maximum_rows=10,
        )
    finally:
        connection.close()
    first = result.loc[result["seed_id"] == "S1"].iloc[0]
    assert "Seed one" in first["associated_seed_protein_names"]
    assert first["seed_protein_names"] == ""
    assert first["annotation_scope"].startswith("cluster-associated")


def test_minimal_seed_authority_retains_raw_identifier_sequence() -> None:
    """An authority-only release works without metadata, links or parsed IDs."""
    connection = duckdb.connect(":memory:")
    try:
        connection.execute("CREATE TABLE known_e3_seeds(seed_id VARCHAR)")
        connection.execute("INSERT INTO known_e3_seeds VALUES ('RAW_SEED')")
        connection.execute(
            "CREATE TABLE candidate_group_member_sequences("
            "raw_identifier VARCHAR, protein_sequence VARCHAR)"
        )
        connection.execute(
            "INSERT INTO candidate_group_member_sequences VALUES "
            "('RAW_SEED', 'MPEPTIDE')"
        )
        result = collect_seed_catalogue(
            connection=connection,
            maximum_rows=10,
        )
    finally:
        connection.close()
    assert result.loc[0, "seed_id"] == "RAW_SEED"
    assert result.loc[0, "seed_protein_names"] == ""
    assert result.loc[0, "source_cluster_count"] == 0
    assert result.loc[0, "sequence_species"] == ""
    assert result.loc[0, "protein_sequence"] == "MPEPTIDE"


def test_seed_capability_skips_incomplete_preferred_relations() -> None:
    """Source selection continues past relations lacking seed identifiers."""
    connection = duckdb.connect(":memory:")
    try:
        connection.execute("CREATE TABLE known_e3_seeds(value VARCHAR)")
        connection.execute("CREATE TABLE candidate_evidence(cluster_id VARCHAR)")
        connection.execute(
            "CREATE TABLE e3_cluster_candidate_evidence("
            "representative_id VARCHAR, known_e3_seed_ids VARCHAR)"
        )
        capability = seed_catalogue_capability(connection=connection)
    finally:
        connection.close()
    assert capability["mode"] == "cluster_summary"
    assert capability["relation"] == "e3_cluster_candidate_evidence"
    assert not capability["sequence_available"]


def test_seed_catalogue_filter_accepts_pasted_terms() -> None:
    """Several literal identifiers or annotations can select catalogue rows."""
    frame = pd.DataFrame(
        data={
            "seed_id": ["S1", "S2", "S3"],
            "associated_seed_protein_names": ["Alpha", "Beta", "Gamma"],
            "protein_sequence": ["MA", "MB", "MC"],
        }
    )
    selected = filter_seed_catalogue(frame=frame, query="S1\nBeta")
    assert selected["seed_id"].tolist() == ["S1", "S2"]
    assert filter_seed_catalogue(frame=frame, query="").equals(frame)


def test_seed_catalogue_validates_source_and_limits() -> None:
    """Unavailable seed evidence and unsafe row bounds fail explicitly."""
    connection = duckdb.connect(":memory:")
    try:
        capability = seed_catalogue_capability(connection=connection)
        assert not capability["available"]
        with pytest.raises(AppError, match="No E3 seed evidence"):
            collect_seed_catalogue(connection=connection)
        capability["available"] = True
        with pytest.raises(AppError, match="between 1 and 100000"):
            build_seed_catalogue_query(
                capability=capability,
                maximum_rows=0,
            )
        with pytest.raises(AppError, match="between 1 and 100000"):
            build_seed_catalogue_query(
                capability=capability,
                maximum_rows=100_001,
            )
    finally:
        connection.close()


def test_seed_catalogue_wraps_duckdb_query_failures(
    seed_connection: duckdb.DuckDBPyConnection,
    monkeypatch: pytest.MonkeyPatch,
) -> None:
    """Runtime SQL errors are translated into the app's controlled exception."""
    monkeypatch.setattr(
        "e3app.seed_catalogue.build_seed_catalogue_query",
        lambda **_: "SELECT * FROM relation_that_does_not_exist",
    )
    with pytest.raises(AppError, match="Could not collect the E3 seed catalogue"):
        collect_seed_catalogue(connection=seed_connection)

File: E3_shiny_app/tests/testthat/test_seed_catalogue.R
test_that("seed catalogue capability requires inherited seed identifiers", {
  capability <- seed_catalogue_capability(list(
    candidate_evidence = c(
      "cluster_id", "matched_seed_ids_calculated", "seed_protein_names"
    ),
    candidate_group_member_sequences = c(
      "parsed_accession", "raw_identifier", "protein_sequence"
    )
  ))
  expect_true(capability$available)
  expect_identical(capability$mode, "cluster_summary")
  expect_identical(capability$relation, "candidate_evidence")
  expect_identical(capability$seed_id_column, "matched_seed_ids_calculated")
  expect_true(capability$sequence_available)
  authority <- seed_catalogue_capability(list(
    known_e3_seeds = c("seed_id", "seed_metadata_json"),
    candidate_evidence = c("cluster_id", "matched_seed_ids_calculated")
  ))
  expect_identical(authority$mode, "authority")
  expect_identical(authority$relation, "known_e3_seeds")
  expect_false(seed_catalogue_capability(list())$available)
})

test_that("seed catalogue query returns annotations and exact sequences", {
  skip_if_not_installed("DBI")
  skip_if_not_installed("duckdb")
  skip_if_not_installed("duckplyr")
  path <- tempfile(fileext = ".duckdb")
  con <- DBI::dbConnect(duckdb::duckdb(), dbdir = path, read_only = FALSE)
  on.exit({
    try(DBI::dbDisconnect(con, shutdown = TRUE), silent = TRUE)
    unlink(path)
  }, add = TRUE)
  DBI::dbWriteTable(
    con,
    "candidate_evidence",
    data.frame(
      cluster_id = c("cluster_1", "cluster_2"),
      matched_seed_ids_calculated = c("S1;S2", "S1"),
      seed_protein_names = c("Seed one;Seed two", "Seed one"),
      seed_categories = c("U-box", "U-box"),
      stringsAsFactors = FALSE
    )
  )
  DBI::dbWriteTable(
    con,
    "known_e3_seeds",
    data.frame(
      seed_id = c("S1", "S2"),
      source_value = c("S1", "S2"),
      source_column = c("accession", "accession"),
      source_row = c(2L, 3L),
      source_path = c("seeds.tsv", "seeds.tsv"),
      seed_metadata_json = c(
        paste0(
          '{"protein_names":"Seed one","e3_category":"U-box",',
          '"organism":"Arabidopsis thaliana","taxon_id":"3702"}'
        ),
        '{"protein_names":"Seed two","category":"BTB"}'
      ),
      stringsAsFactors = FALSE
    )
  )
  DBI::dbWriteTable(
    con,
    "candidate_group_member_sequences",
    data.frame(
      parsed_accession = c("S1", "OTHER"),
      raw_identifier = c("sp|S1|SEED_ONE", "OTHER"),
      species = c("Arabidopsis_thaliana", "Homo_sapiens"),
      protein_sequence = c("MAAA", "MBBB"),
      is_input_candidate = c(FALSE, FALSE),
      stringsAsFactors = FALSE
    )
  )
  DBI::dbDisconnect(con, shutdown = TRUE)
  con <- NULL
  capability <- seed_catalogue_capability(list(
    known_e3_seeds = c(
      "seed_id", "source_value", "source_column", "source_row", "source_path",
      "seed_metadata_json"
    ),
    candidate_evidence = c(
      "cluster_id", "matched_seed_ids_calculated", "seed_protein_names",
      "seed_categories"
    ),
    candidate_group_member_sequences = c(
      "parsed_accession", "raw_identifier", "species", "protein_sequence",
      "is_input_candidate"
    )
  ))
  expect_identical(capability$mode, "authority")
  query <- build_seed_catalogue_query(
    capability = capability,
    max_rows = 10L
  )
  expect_false(grepl("json_extract", query, fixed = TRUE))
  expect_false(grepl("AS JSON", query, fixed = TRUE))
  result <- collect_seed_catalogue(
    duckdb_path = path,
    capability = capability,
    max_rows = 10L
  )
  expect_identical(result$seed_id, c("S1", "S2"))
  expect_equal(result$source_cluster_count, c(2, 1))
  expect_identical(result$seed_protein_names, c("Seed one", "Seed two"))
  expect_identical(result$seed_category, c("U-box", "BTB"))
  expect_identical(result$associated_seed_protein_names, c("", ""))
  expect_identical(result$annotation_scope, rep("exact seed authority row", 2L))
  expect_identical(result$protein_sequence, c("MAAA", ""))
  expect_identical(result$sequence_available, c(TRUE, FALSE))
})

test_that("seed catalogue filters pasted terms and validates bounds", {
  data <- data.frame(
    seed_id = c("S1", "S2", "S3"),
    associated_seed_protein_names = c("Alpha", "Beta", "Gamma"),
    protein_sequence = c("MA", "MB", "MC"),
    stringsAsFactors = FALSE
  )
  expect_identical(
    parse_seed_catalogue_terms(query = "S1\nBeta;NRT"),
    c("s1", "beta", "nrt")
  )
  expect_identical(parse_seed_catalogue_terms(query = NA_character_), character())
  selected <- filter_seed_catalogue(data = data, query = "S1\nBeta")
  expect_identical(selected$seed_id, c("S1", "S2"))
  expect_identical(filter_seed_catalogue(data = data, query = ""), data)
  expect_error(
    parse_seed_catalogue_terms(query = "S1", maximum_terms = 0L),
    "positive integer"
  )
  expect_error(
    build_seed_catalogue_query(
      capability = list(available = TRUE),
      max_rows = 0L
    ),
    "between 1 and 100000"
  )
})

test_that("seed metadata parsing is defensive and preserves existing values", {
  parsed <- parse_seed_catalogue_metadata(
    value = paste0(
      '{"protein_names":"Seed one","e3_category":"U-box",',
      '"reviewed":true,"taxon_id":3702}'
    )
  )
  expect_identical(unname(parsed[["seed_protein_names"]]), "Seed one")
  expect_identical(unname(parsed[["seed_category"]]), "U-box")
  expect_identical(unname(parsed[["seed_review_status"]]), "true")
  expect_identical(unname(parsed[["seed_taxon_id"]]), "3702")
  expect_false(isTRUE(attr(parsed, "parse_failed")))

  invalid <- parse_seed_catalogue_metadata(value = "not-json")
  expect_true(isTRUE(attr(invalid, "parse_failed")))
  frame <- data.frame(
    seed_metadata_json = c("not-json", '{"protein_names":"Parsed"}'),
    seed_protein_names = c("Existing", ""),
    stringsAsFactors = FALSE
  )
  expect_warning(
    enriched <- enrich_seed_catalogue_metadata(data = frame),
    "Ignored 1 invalid"
  )
  expect_identical(enriched$seed_protein_names, c("Existing", "Parsed"))
  expect_error(
    enrich_seed_catalogue_metadata(data = "not-a-data-frame"),
    "requires a data frame"
  )
})

test_that("seed catalogue UI exposes table and three downloads", {
  ui <- as.character(seed_catalogue_ui("seeds"))
  expect_match(ui, "seeds-seed_table", fixed = TRUE)
  expect_match(ui, "seeds-download_tsv", fixed = TRUE)
  expect_match(ui, "seeds-download_excel", fixed = TRUE)
  expect_match(ui, "seeds-fasta_download_ui", fixed = TRUE)
})
